google/cloud/bigtable/backup.py

```diff
@@ -0,0 +1,352 @@
+# from google.cloud.bigtable_admin_v2.proto import bigtable_table_admin_pb2
+from google.cloud.exceptions import NotFound
+
+_BACKUP_NAME_RE = re.compile(
+    r"^projects/(?P<project>[^/]+)/"
+    r"instances/(?P<instance_id>[a-z][-a-z0-9]*)/"
+    r"clusters/(?P<cluster_id>[a-z][-a-z0-9]*)/"
+    r"backups/(?P<backup_id>[a-z][a-z0-9_\-]*[a-z0-9])$"
+)
+_DEFAULT_EXPIRE_PERIOD = 604800 # 7 days
+
+
+class Backup(object):
+	""" Representation of a Google Cloud Bigtable Backup.
+
+	A :class`Backup` can be used to:
+
+	* :meth:`create` the backup
+	* :meth:`update` the backup
+	* :meth:`delete` the backup
+
+	:type backup_id: str
+	:param backup_id: The ID of the backup.
+
+	:type instance: :class:`~google.cloud.spanner_v1.instance.Instance`
+	:param instance: The Instance that owns the Backup.
+
+	:type cluster_id: str
+	:param cluster_id: (Optional) The ID of the Cluster that contains this Backup.
+					   Required for calling 'delete', 'exists' etc. methods.
+
+	:type table_id: str
+	:param table_id: (Optional) The ID of the Table that the Backup is for.
+				  	 Required if the 'create' method will be called.
+
+	:type expire_time: :class:`datetime.datetime`
+	:param expire_time: (Optional) The expiration time after which the Backup
+						will be automatically deleted. Required if the `create`
+						method will be called.
+	"""
+
+	def __init__(self, backup_id, instance,
+				 cluster_id=None, table_id=None, expire_time=None):
+		self.backup_id = backup_id
+		self._instance = instance
+		self._cluster = cluster_id
+		self.table_id = table_id
+		self._expire_time = expire_time
+
+		self._parent = None
+		self._source_table = None
+		self._start_time = None
+		self._end_time = None
+		self._size_bytes = None
+		self._state = None
+
+	@property
+	def name(self):
+		"""Backup name used in requests.
+
+		The Backup name is of the form
+
+			``"projects/../instances/../clusters/../backups/{backup_id}"``
+
+		:rtype: str
+		:returns: The Backup name.
+
+		:raises ValueError: If the 'cluster' has not been set.
+		"""
+		if not self._cluster:
+			raise ValueError('"cluster" parameter must be set')
+
+		return '{}/clusters/{}/backups/{}'.format(
+			self._instance.name, self._cluster, self.backup_id
+		)
+
+	@property
+	def cluster(self):
+		return self._cluster
+
+	@cluster.setter
+	def cluster(self, cluster_id):
+		self._cluster = cluster_id
+
+	@property
+	def parent(self):
+		if not self._parent and self._cluster:
+			self._parent = self._instance.name + '/clusters/' + self._cluster
+		return self._parent
+
+	@property
+	def source_table(self):
+		"""The full name of the Table from which this Backup is created.
+		The table name is of the form
+
+			``"projects/../instances/../tables/{source_table}"``
+
+		:rtype: str
+		:returns: The Table name.
+		"""
+		if not self._source_table and self.table_id:
+			self._source_table = '{}/tables/{}'.format(
+				self._instance.name, self.table_id
+			)
+		return self._source_table
```

Comment:
The same as the comment upper

Reply:
Yep, same idea behind it, as before - basically, an all-or-nothing type of return. I've seen this sort of implementations in Beam. Would it against SoDa conventions, WDYT?